### DISCUSSION TITLE
Mécanisme log

### DISCUSSION THREAD

**Discussion by pbourdu** [publicodes/publicodes#603] [Ideas]
## Use case

Il est parfois utile de pouvoir utiliser la fonction LOG pour des calculs financiers.
Par exemple le calcul du nombre de période dans une simulation de valeur actualisée contenant `n` flux s'obtient avec ce genre de fonction JS

```typescript
export function totalPeriodsCalculation(noOfPeriodsPerYear, currentValue, futureValue, streamValue, rate) {
    let n = 0 == noOfPeriodsPerYear ? 0 : rate / noOfPeriodsPerYear;
    n /= 100;
    let s = rate > 0 ? Math.log((streamValue - futureValue * n) / (streamValue + currentValue * n)) / Math.log(1 + n) : -(futureValue + currentValue) / streamValue;
    return roundToTwoDecimals(s > 0 ? s : 0)
}
```

Aujourd'hui c'est impossible de l'implémenter dans dès règles publicodes.

## Description

La fonction `log()` ne poserait pas de conflit avec l'existant 🤔 ?
Et pour l'implémentation, simplement utiliser `Math.log()` du JS.

```yaml
# Exemple
ma variable:
    valeur: 12

résultat: 
  titre: le log de ma variable
  valeur: log(variable)
  arrondi: 2 décimales
```

Qu'en dis tu @johangirod ?

**Comment by johangirod** [comment]
Ça me semble une très bonne idée comme ajout de fonctionalité.

Ceci étant dit, la syntaxe proposée introduit une notion de taille, qui n'existe pas actuellement : la possibilité d'appeler des fonctions (ou mécanisme) directement dans une expression publicodes `fn(x)`

Je serai d'avis d'avoir une première version qui utilise la syntaxe de mécanisme existant sans en ajouter de nouvelle : 

```
résultat: 
  titre: le log de ma variable
  logarithme: variable
  arrondi: 2 décimales
```

**Comment by pbourdu** [comment]
Merci pour ton retour @johangirod 
Je suis bien tenté d'essayer l'implémentation.
J'ai lu [la doc](https://github.com/publicodes/publicodes/blob/master/CONTRIBUTING.md#ajouter-un-nouveau-m%C3%A9canisme), j'ai un peu de mal à situer les modification à apporter.

J'ai établi cette liste de course : 
1. Ajouter un fichier `logarithme.ts` dans https://github.com/publicodes/publicodes/tree/master/packages/core/src/mecanisms qui pour l'instant ressemble à ça 😅. Ça me soulève pas mal de questions que je mets direct en commentaire dedans
```typescript
import { EvaluationFunction } from '..'
import { registerEvaluationFunction } from '../evaluationFunctions'

export type LogarithmeNode = {
    nodeKind: 'logarithme',
    explanation: {
        // ou est défini let objet ?
    }
}

export default function parseLogarithme(v, _context) {
    const explanation = {
        // ??
    }
    return {
        nodeKind: parseLogarithme.nom,
        explanation,
    }
}

parseLogarithme.nom = 'logarithme' as const

const evaluate: EvaluationFunction<'logarithme'> = function(node: LogarithmeNode) { // est ce bien ce type d'objet qui est passé
    // À faire ??
    // Quel format format de retour
}

registerEvaluationFunction(parseLogarithme.nom, evaluate)

```

2. Un fichier de tests `logarithme.yaml` dans https://github.com/publicodes/publicodes/tree/master/packages/core/test/m%C3%A9canismes
```yaml
logarithme test:
  logarithme: 1
  exemples:
    - nom: 'log de 1 = 0'
      valeur attendue: 0

```

3. Le référencement dans https://github.com/publicodes/publicodes/blob/master/packages/core/src/parse.ts, dans quel partie ? Mécanisme chaîné ?

Pour l'instant je n'arrive pas à toucher ma fonction de parsing avec un point d'arrêt sur `parseLogarithme` en mode debug 😬 
Aussi preneur de ressources pour comprendre un peu mieux l'organisation du package core.
Par avance merci de ton aide.

**Reply by johangirod** [reply]
Top, c'est un super bon début ! 

1. La clé `explanation` contient le reste de l'arbre sur lequel est appliqué le logarithme (comme c'est un mécanisme unaire, il n'y a qu'un seul enfant)
```ts
 type LogarithmeNode = {
  explanation: ASTNode
  nodeKind: 'logarithme'
}
```

2. L'AST est formé en appelant recursivement `parse` sur le noeud enfant
```ts
export default function parseLogarithme(v, _context) {
    const explanation = parse(v)
    return {
        nodeKind: parseLogarithme.nom,
        explanation,
    }
}
```
4. Pour l'évaluation, la fonction doit retourner un objet de type [EvaluatedNode](https://publi.codes/docs/api/publicodes/type-aliases/evaluatednode). Pour cela il faudra évaluer récursivement le noeud enfant (dans evaluation), et calculer le logarithme à partir de la valeur retournée. 
5. La question de l'unité est à traiter : quelle est l'unité de `log(4 ans)` ? Doit-on autoriser de calculer un logarithme sur des nombres avec unité ou non ?

6. A priori, ce ne sera pas un mécanisme chainé (comme plafond ou plancher) mais un mécanisme classique. Il faut donc l'ajouter uniquement dans `parseFunction`

7. Pour terminer, il faut également ajouter une logique pour le visiteur de l'AST : 
  - En listant le mécanisme comme mécanisme unaire : https://github.com/publicodes/publicodes/blob/master/packages/core/src/AST/index.ts#L142
  - En ajoutant le type du noeud : https://github.com/publicodes/publicodes/blob/master/packages/core/src/AST/types.ts#L31
  - En spécifiant l'inférence de type (ici nombre, comme pour le mécanisme unité) : https://github.com/publicodes/publicodes/blob/master/packages/core/src/inferNodeType.ts#L88

**Comment by pbourdu** [comment]
Cette fonctionnalité a été implémenté dans https://github.com/publicodes/publicodes/releases/tag/publicodes%401.7.0, je close ce thread.
